File: Windows/socks5messagequeue.cpp
/*
 * Title:  spider socks5messagequeue.cpp (Windows)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "socks5messagequeue.hpp"
#include "socks5message.hpp"

namespace spider
{
    Socks5messagequeue::Socks5messagequeue()
    {

    }

    Socks5messagequeue::~Socks5messagequeue()
    {

    }

    void Socks5messagequeue::push(std::shared_ptr<Socks5message> message)
    {
        guard.acquire();
        queue.push(message);
        guard.release();

        token.release();
    }

    int32_t Socks5messagequeue::push_timeout(std::shared_ptr<Socks5message> message,
                                             int32_t tv_sec,
                                             int32_t tv_usec)
    {
        int32_t ret = 0;
        int32_t t = tv_sec * 1000000 + tv_usec;
        auto duration = std::chrono::microseconds{t};

        if(guard.try_acquire_for(duration))
        {
            queue.push(message);
            guard.release();
            token.release();
        }else
        {
#ifdef DEBUGPRINT
            std::printf("[-] push timeout\n");
#endif
            ret = -1;
        }

        return ret;
    }

    std::shared_ptr<Socks5message> Socks5messagequeue::pop()
    {
        token.acquire();

        guard.acquire();
        std::shared_ptr<Socks5message> message = queue.front();
        queue.pop();
        guard.release();

        return message;
    }

    std::shared_ptr<Socks5message> Socks5messagequeue::pop_timeout(int32_t tv_sec,
                                                                   int32_t tv_usec)
    {
        std::shared_ptr<Socks5message> message = nullptr;
        int32_t t = tv_sec * 1000000 + tv_usec;
        auto duration = std::chrono::microseconds{t};

        if(token.try_acquire_for(duration))
        {
            guard.acquire();
            message = queue.front();
            queue.pop();
            guard.release();
        }else
        {
#ifdef DEBUGPRINT
            std::printf("[-] pop timeout\n");
#endif
        }

        return message;
    }
}

File: Windows_csharp/Socks5MessageQueue.cs
/*
 * Title:  spider Socks5MessageQueue.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Collections.Generic;
using System.Threading;

namespace spider
{
    public class Socks5MessageQueue
    {
        private const int SOCKS5_MESSAGE_QUEUE_CAPACITY = 100;
        private Queue<Socks5Message> queue = new Queue<Socks5Message>();
        private SemaphoreSlim token = new SemaphoreSlim(0, SOCKS5_MESSAGE_QUEUE_CAPACITY);
        private SemaphoreSlim guard = new SemaphoreSlim(1, 1);
        private int count = 0;

        public Socks5MessageQueue()
        {

        }

        ~Socks5MessageQueue()
        {

        }

        public int GetCount()
        {
            return count;
        }

        public void Push(Socks5Message message)
        {
            guard.Wait();
            try
            {
                while(token.CurrentCount >= SOCKS5_MESSAGE_QUEUE_CAPACITY)
                {
                    guard.Release();
                    Thread.Sleep(10);
                    guard.Wait();
                }
                queue.Enqueue(message);
                count++;
                token.Release();
            }finally
            {
                guard.Release();
            }
        }

        public int PushTimeout(Socks5Message message,
                               int tv_sec,
                               int tv_usec)
        {
            int ret = 0;
            int t = tv_sec * 1000 + tv_usec / 1000; // milliseconds

            if(guard.Wait(t))
            {
                try
                {
                    while(token.CurrentCount >= SOCKS5_MESSAGE_QUEUE_CAPACITY)
                    {
                        guard.Release();
                        Thread.Sleep(10);
                        guard.Wait();
                    }
                    queue.Enqueue(message);
                    count++;
                    token.Release();
                }finally
                {
                    guard.Release();
                }
            }else
            {
#if DEBUGPRINT
                Console.WriteLine("[-] push timeout");
#endif
                ret = -1;
            }

            return ret;
        }

        public Socks5Message Pop()
        {
            Socks5Message message;

            token.Wait();
            guard.Wait();
            message = queue.Dequeue();
            count--;
            guard.Release();

            return message;
        }

        public Socks5Message PopTimeout(int tv_sec,
                                        int tv_usec)
        {
            Socks5Message message = null;
            int t = tv_sec * 1000 + tv_usec / 1000; // milliseconds

            if(token.Wait(t))
            {
                guard.Wait();
                message = queue.Dequeue();
                count--;
                guard.Release();
            }else
            {
#if DEBUGPRINT
                Console.WriteLine("[-] pop timeout");
#endif
            }

            return message;
        }
    }
}

